type annotation-driven cleanup

williballenthin opened the pull request:
This PR makes a lot of changes derived from type annotations and other warnings in VS Code/pylance. I hope that some or all of the changes are acceptable, and I'm totally open to rolling some of them back based on your feedback. I think many of the changes are worthwhile, since they substantially reduce the number of highlights and warnings displayed in my IDEs :-)

Functionally, only a few things have changed in small ways. Most of the changes tweak the types of variables/routines or adjust how classes are used to store data (particularly, by avoiding class variables when possible).

Here are some of the notable changes:

### default values removed for class variables
Almost all class variables are used just to provide type hints for `Structure` subclasses. I've removed the default values (typically `None`) from their declarations, which makes type checkers much happier. The hints still work for instances, but no variables are actually created on the class. 

![image](https://user-images.githubusercontent.com/156560/145648982-51862f01-b5e4-4a6d-a55f-80758fcaaf18.png)

However, this does mean you have to use `if hasattr(it, "foo")` rather than `if it.foo`, though this only affects internal code:
![image](https://user-images.githubusercontent.com/156560/145649248-95d1aae1-52c1-48cc-9724-d72edc41955d.png)

### removes class variables when possible
For `MDTableRow`, `ClrMetaDataTable`, and friends, move class variables into instance variables where appropriate. As far as I can tell, class variables are needed for the dynamically parsed instances (like `MethodRowStruct`) but not in some other cases. As I'm sure you know, class variables can be a little tricky, so I'd argue that instance variables should be preferred when possible. As always, open to feedback and discussion.

With the above in mind, I've made the class variable accesses extra explicit by routing them through `self.__class__.foo` rather than `self.foo`. This is stylistic and can be reverted.

![image](https://user-images.githubusercontent.com/156560/145649362-200a2782-e5c6-4bad-ba34-038ee430e783.png)

### row type annotations
Introduced type annotations so that row types can be accurately modeled. This means that editors can auto-complete the properties of rows based on each table! Each table uses a generic type parameter to specify the row type.

![image](https://user-images.githubusercontent.com/156560/145649907-ed38f927-1b25-41eb-aab8-c3c747e13860.png)
![image](https://user-images.githubusercontent.com/156560/145650023-dcd0bd66-92e6-4580-810c-c161470f401f.png)

### row field enums
Add support for decode enum data from row properties:

![image](https://user-images.githubusercontent.com/156560/145655494-3c389fea-31b1-4ab5-acd4-a51d72d9f3bc.png)

### mypyp in CI
Adds a CI lint for mypy type checking:
![image](https://user-images.githubusercontent.com/156560/145655526-6e252509-a664-42d7-9a16-8fa245b36253.png)

### adds user_strings heap shortcut
Like blobs and guids, add a shortcut to access user strings:
![image](https://user-images.githubusercontent.com/156560/145902794-b9aa89a5-13ae-4cc0-b635-79fc6cc12cb9.png)

williballenthin commented:
didn't think I'd get mypy passing but only one issue left, so going to try to knock that out, one sec.

williballenthin commented:
![image](https://user-images.githubusercontent.com/156560/145655580-8de51046-58e9-4d36-ba34-a53a29982a5a.png)
:-)

malwarefrank commented:
I mostly like these changes. The two that I have to think more about are whether stream names are guaranteed to be ascii and raising exceptions.

I am not sure if stream names are requiremed to be ascii for a .NET program to load and run.  If it is not a runtime dependency, then we need to account for when the stream name contains non-ascii bytes.

In general I am ok with raising exceptions. However if raising an exception at an early stage or parsing prevents parsing data that would be parsable, then I prefer not to raise an exception. One of my goals in designing dnfile was to parse as much information from the file as possible, even if it is malformed. I believe the pefile mechanism for handling that is to have "warnings" that are shown when print_info() is called. I have tried to use that mechanism, though my current implementation is imperfect.

williballenthin commented:
Great feedback, thanks for that!

Tomorrow I’ll do at least two things:

  - Research the stream name encoding, and provide a test case either way. If the names can be arbitrary bytes then I completely agree they should be represented this way in dnfile. I’ll look at the spec and then see how the M$ and mono runtimes handle malformed data. 

  - enumerate the places I changed the exception behavior. Off the top of my head I *think* it’s basically two cases:

    - top level can’t parse any metadata (previous behavior was just to set pe.net to None)
    - inner routines, like pick struct format, whose error mechanism should simply become more descriptive with exceptions, not change the functionality. 

I might have forgotten some here, and regardless, happy to revert if necessary. I agree that the library should parse as much as necessary, so please highlight any areas I can do a better job. 


> On Dec 12, 2021, at 6:59 PM, malwarefrank ***@***.***> wrote:
> 
> ﻿
> I mostly like these changes. The two that I have to think more about are whether stream names are guaranteed to be ascii and raising exceptions.
> 
> I am not sure if stream names are requiremed to be ascii for a .NET program to load and run. If it is not a runtime dependency, then we need to account for when the stream name contains non-ascii bytes.
> 
> In general I am ok with raising exceptions. However if raising an exception at an early stage or parsing prevents parsing data that would be parsable, then I prefer not to raise an exception. One of my goals in designing dnfile was to parse as much information from the file as possible, even if it is malformed. I believe the pefile mechanism for handling that is to have "warnings" that are shown when print_info() is called. I have tried to use that mechanism, though my current implementation is imperfect.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

williballenthin commented:
duplicate stream test added here: https://github.com/malwarefrank/dnfile/pull/17/commits/614579c38bea10798d285ec2608650d36369020a

williballenthin commented:
research into stream name handling is in #19. in summary, there can be duplicate streams (last one wins) and non-ASCII stream names. 

therefore, I've reverted the stream-name-as-ASCII changes and added tests to demonstrate the handling. this will required https://github.com/malwarefrank/dnfile-testfiles/pull/2 to be merged and the submodule updated here. passing locally for me:

![image](https://user-images.githubusercontent.com/156560/145873805-65f50d2e-0700-427a-8cca-917e17f3a8ab.png)

williballenthin commented:
I've reverted the changes that introduce exceptions. I should have done a much better job preserving behavior of the library when adding them before. Now, I've gone through all the changes involving exceptions and validated that:
  1. they are required, such as in `ClrData.__init__` and there's no other option, or
  2. they are even needed, such as in `clr_coded_index_struct_size` where the checks weren't even required, or
  3. they have been replaced with returning None, such as in `BinaryHeap.get`

For (3), which was the majority, I updated the type signatures to indicate the return value is `Optional[T]` which lets mypy identify when the caller doesn't check for failure. I've always liked exceptions for errors, but with mypy able to identify this, I actually feel more confident that we handle the error cases more consistently (since python/mypy doesn't have anything like "checked exceptions").

Please let me know what you think. I don't have any remaining outstanding changes.

williballenthin commented on examples/dndump.py:
```diff
@@ -188,7 +191,11 @@ def render_pe(ostream: Formatter, dn):
-                                    name = None if not hasattr(v, "table") else v.table.name
+                                    if not hasattr(v, "table") or v.table is None:
+                                        logger.warning("reference has no table: %s", v)
+                                        name  = "(missing)"
```
```suggestion
                                        name = "(missing)"
```

williballenthin commented on examples/dndump.py:
```diff
@@ -232,7 +239,12 @@ def render_pe(ostream: Formatter, dn):
-                                            name = None if not hasattr(vv, "table") else vv.table.name
+                                            if not hasattr(vv, "table") or vv.table is None:
+                                                logger.warning("reference has no table: %s", vv)
+                                                name  = "(missing)"
```
```suggestion
                                                name = "(missing)"
```

williballenthin commented on src/dnfile/__init__.py:
```diff
@@ -338,20 +357,23 @@ def parse_stream_table(self, pe: dnPE, streams_table_rva):
+                # if a stream with this name already exists.
+                # this is not fatal, just unusual.
+                pe.add_warning("Duplicate .NET stream name '{!r}'".format(name))
+                logger.warning("Duplicate .NET stream with name: %s", name)
```
```suggestion
                logger.warning("Duplicate .NET stream name: %s", name)
```

malwarefrank commented on src/dnfile/codedindex.py:
```diff
@@ -12,26 +12,55 @@
 from .base import CodedIndex
 
+if TYPE_CHECKING:
+    from .mdtable import *
```
I am generally not a fan of "import *".  I want to get these changes in sooner rather than later, so I may change that post-merge

williballenthin replied:
no problem. would you prefer keeping the namespace (e.g. `import .mdtable as mdtable` or `import dnfile.mdtable`) or using bare symbols (e.g. `from .mdtable import Foo, Bar, Baz, ...`)? i'll start with the former.

williballenthin replied:
fixed in https://github.com/malwarefrank/dnfile/pull/17/commits/11da6b592b70a93fc9e3189aa7e118a53baeb31b

malwarefrank commented on tests/test_parse.py:
```diff
@@ -124,4 +136,14 @@ def test_method_params():
+
+def test_ignore_NumberOfRvaAndSizes():
+    # .NET loaders ignores NumberOfRvaAndSizes, so attempt to parse anyways
+    path = fixtures.DATA / "1d41308bf4148b4c138f9307abc696a6e4c05a5a89ddeb8926317685abb1c241"
```
Looks like a malware test file / hash made its way into the tests.  I am ok with having malware examples, but not distributing in testfiles, so may make those tests optional for people who do not have the data

williballenthin replied:
ah, ok, that's a fair policy. i'll update the test to check for file existence, first

williballenthin replied:
fixed in https://github.com/malwarefrank/dnfile/pull/17/commits/877b4caca913537985f4e8dd5ea8805b93350fa9

malwarefrank commented:
wow, you really did a lot of work here, thanks!  I want to tag the current main branch and then will merge this in

williballenthin commented:
tests looking good, i have no pending changes. thanks for the reviews!

williballenthin commented:
(unrelated...)

we have four exceptions tracked internally after running dnfile against a bunch of malware. i'll likely be digging into those tomorrow and reporting the findings to this repo. i don't anticipate remotely the same amount of code as this PR.

after that, probably two things to look forward two:
  1. further stress testing against additional malware samples, and
  2. (probably) starting to build a project on dnfil, and will report any feature requests/enhancement here.

feel free to reach out if you'd prefer any coordination. so far, my colleagues and I have enjoyed working with you and really appreciate your attention to our issues and PRs!
